Sanitize IPsec key after setting config

ffoulkes opened the pull request:
- Implement wrapper class to zero key field in sadb_config object.

- Use clang-format to fix line wrapping and indentation.

ffoulkes committed:
Zero memory after setting IPsec key

- Implement wrapper class to zero key field in sadb_config object.

- Use clang-format to fix line wrapping and indentation.

Signed-off-by: Derek G Foster <derek.foster@intel.com>

ffoulkes commented:
I get the following compiler warning when I run the Stratum build:
```text
dfoster@dgfoster-mobl1:~/latest/stratum/stratum$ bazel build //stratum/hal/bin/tdi/es2k:stratum_es2k --define target=es2k
Starting local Bazel server and connecting to it...
DEBUG: /home/dfoster/latest/stratum/stratum/stratum/hal/lib/tdi/es2k/es2k.bzl:18:10: Target is: 'ES2K'
INFO: Analyzed target //stratum/hal/bin/tdi/es2k:stratum_es2k (225 packages loaded, 12692 targets configured).
INFO: Found 1 target...
INFO: From Compiling stratum/hal/lib/tdi/tdi_fixed_function_manager.cc:
stratum/hal/lib/tdi/tdi_fixed_function_manager.cc: In destructor 'stratum::hal::tdi::SadbConfigWrapper::~SadbConfigWrapper()':
stratum/hal/lib/tdi/tdi_fixed_function_manager.cc:25:41: warning: 'void* memset(void*, int, size_t)' clearing an object of type 'std::string' {aka 'class std::__cxx11::basic_string<char>'} with no trivial copy-assignment; use assignment or value-initialization instead [-Wclass-memaccess]
   25 |       memset(&key[0], 0, key->capacity());
      |                                         ^
In file included from /usr/include/c++/9/string:55,
                 from external/com_google_absl/absl/synchronization/mutex.h:63,
                 from ./stratum/hal/lib/tdi/tdi_fixed_function_manager.h:7,
                 from stratum/hal/lib/tdi/tdi_fixed_function_manager.cc:4:
/usr/include/c++/9/bits/basic_string.h:81:11: note: 'std::string' {aka 'class std::__cxx11::basic_string<char>'} declared here
   81 |     class basic_string
      |           ^~~~~~~~~~~~
Target //stratum/hal/bin/tdi/es2k:stratum_es2k up-to-date:
  bazel-bin/stratum/hal/bin/tdi/es2k/stratum_es2k
INFO: Elapsed time: 17.557s, Critical Path: 6.73s
INFO: 3 processes: 1 internal, 2 linux-sandbox.
INFO: Build completed successfully, 3 total actions
```
Any suggestions?

nupuruttarwar commented on stratum/hal/lib/tdi/tdi_fixed_function_manager.cc:
```diff
@@ -74,70 +99,71 @@ ::util::Status TdiFixedFunctionManager::WriteSadbEntry(
-    RETURN_IF_ERROR(table_data->SetParam(kIpsecSadbSpi, sadb_config.spi()));
+    TdiSdeInterface::TableDataInterface* table_data,
+    IPsecSADBConfig& sadb_config) {
+  // Wrapper object ensures that sadb_config is scrubbed on exit.
```
Can we say IPsec key is zeroed instead of scrubbed?

ffoulkes replied:
It doesn't matter what we write to the location. I didn't want the abstract interface to be tied to the value used.
I will substitute the term "sanitized", which is used to describe this operation in other contexts.

nupuruttarwar replied:
Okay makes sense, please change the title as well.

nupuruttarwar commented on stratum/hal/lib/tdi/tdi_fixed_function_manager.cc:
```diff
@@ -74,70 +99,71 @@ ::util::Status TdiFixedFunctionManager::WriteSadbEntry(
+    TdiSdeInterface::TableDataInterface* table_data,
+    IPsecSADBConfig& sadb_config) {
+  // Wrapper object ensures that sadb_config is scrubbed on exit.
+  SadbConfigWrapper w(sadb_config);
```
Single-letter names are discouraged. I understand here `w` is not used in the function, but please specify a meaningful name if possible.

ffoulkes replied:
I modeled this line of code on the following idiom, which occurs throughout Stratum:
```c
absl::WriterMutexLock l(&data_lock_);
```
The variable is never referenced. It exists only so we can harness its destructor. I was concerned that giving it a "real" name would lead the reader to assume it was significant in and of itself.

But it occurs to me that I can kill two birds with one stone:
```c
SadbConfigWrapper sanitize_on_exit(sadb_config);
```
and delete the preceding comment line.

nupuruttarwar reviewed (approved):
Minor comments, logically LGTM

nupuruttarwar commented:
> I get the following compiler warning when I run the Stratum build:
> 
> ```
> dfoster@dgfoster-mobl1:~/latest/stratum/stratum$ bazel build //stratum/hal/bin/tdi/es2k:stratum_es2k --define target=es2k
> Starting local Bazel server and connecting to it...
> DEBUG: /home/dfoster/latest/stratum/stratum/stratum/hal/lib/tdi/es2k/es2k.bzl:18:10: Target is: 'ES2K'
> INFO: Analyzed target //stratum/hal/bin/tdi/es2k:stratum_es2k (225 packages loaded, 12692 targets configured).
> INFO: Found 1 target...
> INFO: From Compiling stratum/hal/lib/tdi/tdi_fixed_function_manager.cc:
> stratum/hal/lib/tdi/tdi_fixed_function_manager.cc: In destructor 'stratum::hal::tdi::SadbConfigWrapper::~SadbConfigWrapper()':
> stratum/hal/lib/tdi/tdi_fixed_function_manager.cc:25:41: warning: 'void* memset(void*, int, size_t)' clearing an object of type 'std::string' {aka 'class std::__cxx11::basic_string<char>'} with no trivial copy-assignment; use assignment or value-initialization instead [-Wclass-memaccess]
>    25 |       memset(&key[0], 0, key->capacity());
>       |                                         ^
> In file included from /usr/include/c++/9/string:55,
>                  from external/com_google_absl/absl/synchronization/mutex.h:63,
>                  from ./stratum/hal/lib/tdi/tdi_fixed_function_manager.h:7,
>                  from stratum/hal/lib/tdi/tdi_fixed_function_manager.cc:4:
> /usr/include/c++/9/bits/basic_string.h:81:11: note: 'std::string' {aka 'class std::__cxx11::basic_string<char>'} declared here
>    81 |     class basic_string
>       |           ^~~~~~~~~~~~
> Target //stratum/hal/bin/tdi/es2k:stratum_es2k up-to-date:
>   bazel-bin/stratum/hal/bin/tdi/es2k/stratum_es2k
> INFO: Elapsed time: 17.557s, Critical Path: 6.73s
> INFO: 3 processes: 1 internal, 2 linux-sandbox.
> INFO: Build completed successfully, 3 total actions
> ```
> 
> Any suggestions?



> I get the following compiler warning when I run the Stratum build:
> 
> ```
> dfoster@dgfoster-mobl1:~/latest/stratum/stratum$ bazel build //stratum/hal/bin/tdi/es2k:stratum_es2k --define target=es2k
> Starting local Bazel server and connecting to it...
> DEBUG: /home/dfoster/latest/stratum/stratum/stratum/hal/lib/tdi/es2k/es2k.bzl:18:10: Target is: 'ES2K'
> INFO: Analyzed target //stratum/hal/bin/tdi/es2k:stratum_es2k (225 packages loaded, 12692 targets configured).
> INFO: Found 1 target...
> INFO: From Compiling stratum/hal/lib/tdi/tdi_fixed_function_manager.cc:
> stratum/hal/lib/tdi/tdi_fixed_function_manager.cc: In destructor 'stratum::hal::tdi::SadbConfigWrapper::~SadbConfigWrapper()':
> stratum/hal/lib/tdi/tdi_fixed_function_manager.cc:25:41: warning: 'void* memset(void*, int, size_t)' clearing an object of type 'std::string' {aka 'class std::__cxx11::basic_string<char>'} with no trivial copy-assignment; use assignment or value-initialization instead [-Wclass-memaccess]
>    25 |       memset(&key[0], 0, key->capacity());
>       |                                         ^
> In file included from /usr/include/c++/9/string:55,
>                  from external/com_google_absl/absl/synchronization/mutex.h:63,
>                  from ./stratum/hal/lib/tdi/tdi_fixed_function_manager.h:7,
>                  from stratum/hal/lib/tdi/tdi_fixed_function_manager.cc:4:
> /usr/include/c++/9/bits/basic_string.h:81:11: note: 'std::string' {aka 'class std::__cxx11::basic_string<char>'} declared here
>    81 |     class basic_string
>       |           ^~~~~~~~~~~~
> Target //stratum/hal/bin/tdi/es2k:stratum_es2k up-to-date:
>   bazel-bin/stratum/hal/bin/tdi/es2k/stratum_es2k
> INFO: Elapsed time: 17.557s, Critical Path: 6.73s
> INFO: 3 processes: 1 internal, 2 linux-sandbox.
> INFO: Build completed successfully, 3 total actions
> ```
> 
> Any suggestions?

Referring https://stackoverflow.com/a/62213067

Maybe key doesn't have a [trivial copy assignment operator](https://en.cppreference.com/w/cpp/language/copy_assignment#Trivial_copy_assignment_operator). You can't use memset.
The suggestion was to remove the constructor or use a direct assignment. Can you try direct assignment instead?

ffoulkes commented:
If I feed the following to Compiler Explorer:
```c
#include <memory.h>
#include <string>

class IPsecSADBConfig {
 public:
  std::string key;
};

class SadbConfigWrapper {
 public:
  explicit SadbConfigWrapper(IPsecSADBConfig& sadb_config)
      : config_(sadb_config) {}
  ~SadbConfigWrapper() {
    std::string* key = get_key_ptr();
    if (key) {
      memset(&key[0], 0, key->capacity());
    }
  }

 private:
  // Returns mutable pointer to key field.
  inline std::string* get_key_ptr() {
    return &config_.key;
  }
  IPsecSADBConfig& config_;
};

void dummy(IPsecSADBConfig& sadb_config) {
    SadbConfigWrapper w(sadb_config);
}
```
The destructor looks like this:
```s
SadbConfigWrapper::~SadbConfigWrapper() [base object destructor]:
        push    rbp
        mov     rbp, rsp
        sub     rsp, 32
        mov     QWORD PTR [rbp-24], rdi
        mov     rax, QWORD PTR [rbp-24]
        mov     rdi, rax
        call    SadbConfigWrapper::get_key_ptr[abi:cxx11]()
        mov     QWORD PTR [rbp-8], rax
        cmp     QWORD PTR [rbp-8], 0
        je      .L4
        mov     rax, QWORD PTR [rbp-8]
        mov     rdi, rax
        call    std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::capacity() const
        mov     rdx, rax
        mov     rax, QWORD PTR [rbp-8]
        mov     esi, 0
        mov     rdi, rax
        call    memset
.L4:
        nop
        leave
        ret
```
Even at `-O3` (with the compiler inlining everything), the code sample still calls `memset`:
```text
dummy(IPsecSADBConfig&):             # @dummy(IPsecSADBConfig&)
        mov     rax, rdi
        add     rdi, 8
        lea     rcx, [rax + 24]
        cmp     qword ptr [rax + 8], rcx
        je      .LBB0_1
        mov     rdx, qword ptr [rax + 24]
        mov     esi, 255
        jmp     memset                  # TAILCALL
.LBB0_1:
        mov     edx, 15
        mov     esi, 255
        jmp     memset                  # TAILCALL
```
I added a couple of extra fields SadbConfigWrapper and changed the sanitization value to 0xff for this iteration.

ffoulkes commented:
> Referring https://stackoverflow.com/a/62213067

> Maybe key doesn't have a [trivial copy assignment operator](https://en.cppreference.com/w/cpp/language/copy_assignment#Trivial_copy_assignment_operator). You can't use memset.
The suggestion was to remove the constructor or use a direct assignment. Can you try direct assignment instead?

- `key` is a std::string. I assume it has all the requisite methods.
- `IPsecSADBConfig` is a protobuf object. it's chock-full of fields.
- We need to sanitize just the key, not the entire object.
- We need to sanitize the entire payload of the key.
- We need to sanitize the payload in place (same memory).
- The size of the payload depends on what was specified in the gNMI request.

I don't see how we can use direct assignment to accomplish this.

ffoulkes commented:
I can make the warning go away by doing this:
```c
#pragma GCC diagnostic push
#pragma GCC diagnostic ignored "-Wclass-memaccess"
      memset(&key[0], 0, key->capacity());
#pragma GCC diagnostic pop
```

ffoulkes committed:
Update following review

- Rename wrapper variable from 'w' to 'sanitize_on_exit`.

- Add pragmas to suppress compiler diagnostic on memset() call.

Signed-off-by: Derek G Foster <derek.foster@intel.com>

ffoulkes commented:
For the record:
```text
dfoster@dgfoster-mobl1:~/latest/stratum/stratum$ bazel test //stratum/hal/lib/tdi:all --define target=es2k
DEBUG: /home/dfoster/latest/stratum/stratum/stratum/hal/lib/tdi/es2k/es2k.bzl:18:10: Target is: 'ES2K'
INFO: Analyzed 34 targets (0 packages loaded, 20 targets configured).
INFO: Found 26 targets and 8 test targets...
INFO: Elapsed time: 0.715s, Critical Path: 0.18s
INFO: 2 processes: 1 internal, 1 linux-sandbox.
INFO: Build completed successfully, 2 total actions
//stratum/hal/lib/tdi:tdi_action_profile_manager_test           (cached) PASSED in 0.9s
//stratum/hal/lib/tdi:tdi_counter_manager_test                  (cached) PASSED in 0.8s
//stratum/hal/lib/tdi:tdi_fixed_function_manager_test           (cached) PASSED in 0.2s
//stratum/hal/lib/tdi:tdi_packetio_manager_test                 (cached) PASSED in 1.0s
//stratum/hal/lib/tdi:tdi_pipeline_utils_test                   (cached) PASSED in 0.7s
//stratum/hal/lib/tdi:tdi_pre_manager_test                      (cached) PASSED in 0.3s
//stratum/hal/lib/tdi:tdi_table_manager_test                    (cached) PASSED in 0.9s
//stratum/hal/lib/tdi:utils_test                                (cached) PASSED in 0.8s

INFO: Build completed successfully, 2 total actions
```

ffoulkes committed:
Use OPENSSL_cleanse() to sanitize key string

Signed-off-by: Derek G Foster <derek.foster@intel.com>

5abeel reviewed (approved):
Verified testing this branch that it works. 

LGTM
